Python/chapter03/p01_three_in_one/miguelHx.py

```diff
@@ -0,0 +1,420 @@
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        self.size -= 1
+        return item
+    
+    def push(self, item: int) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item)
+        t.next = self.top
+        self.top = t
+        self.size += 1
+
+    def peek(self) -> int:
+        """
+        Returns data at the top of the stack
+
+        Raises:
+            IndexError: [description]
+
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+    
+    def is_empty(self) -> bool:
+        """
+        Returns True if and only if the
+        stack is empty, otherwise False.
+
+        Returns:
+            bool: True if empty, False otherwise
+        """
+        return self.size == 0
+    
+    def as_list(self) -> List[int]:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(n.data)
+            n = n.next
+        values.append(n.data)
+        return values
+    
+    def __str__(self):
+        if self.size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+class StackTrio(object):
+
+    def __init__(self, stack_capacity = 7):
+        self.num_stacks = 3
+        self.stack_capacity = stack_capacity
+        self.stack_info = {
+            1: {
+                'start': 0,
+                'end': stack_capacity - 1,
+                'size': 0,
+                'top_index': 0,
+                'top_index_next': -1, # index after top, initially starts at a sentinel value because the stack is empty, so there is no "next"
+            },
+            2: {
+                'start': stack_capacity,
+                'end': stack_capacity * 2 - 1,
+                'size': 0,
+                'top_index': stack_capacity,
+                'top_index_next': -1,
+            },
+            3: {
+                'start': stack_capacity * 2,
+                'end': stack_capacity * 3 - 1,
+                'size': 0,
+                'top_index': stack_capacity * 2,
+                'top_index_next': -1
+            }
+        }
+        self.values = [0] * (stack_capacity * self.num_stacks)
+
+    def is_empty(self, stack_id):
+        return self.stack_info[stack_id]['size'] <= 0
```

Comment:
what happens if a caller provides a `stack_id` that's not in the range (1...3)?

Reply:
Good catch 👍  fixed in cb1fe7577bd3dbfc8ea3f6ec4915e3674a3f11c9